core/predicate_check_test.go

```diff
@@ -322,3 +320,141 @@ func TestCheckPredicate(t *testing.T) {
+			testTuple: []testTuple{
+				{address: addr1, isValidPredicate: true},
+				{address: addr1, isValidPredicate: false},
+				{address: addr1, isValidPredicate: false},
+				{address: addr1, isValidPredicate: true},
+			},
+			expectedRes: map[common.Address][]byte{addr1: {6}},
+		},
+		{
+			name: "two addresses mixed predicates",
+			testTuple: []testTuple{
+				{address: addr1, isValidPredicate: true},
+				{address: addr2, isValidPredicate: false},
+				{address: addr1, isValidPredicate: false},
+				{address: addr1, isValidPredicate: false},
+				{address: addr2, isValidPredicate: true},
+				{address: addr2, isValidPredicate: true},
+				{address: addr2, isValidPredicate: false},
+				{address: addr2, isValidPredicate: true},
+			},
+			expectedRes: map[common.Address][]byte{addr1: {6}, addr2: {9}},
+		},
+		{
+			name: "two addresses all valid predicates",
+			testTuple: []testTuple{
+				{address: addr1, isValidPredicate: true},
+				{address: addr2, isValidPredicate: true},
+				{address: addr1, isValidPredicate: true},
+				{address: addr1, isValidPredicate: true},
+			},
+			expectedRes: map[common.Address][]byte{addr1: {}, addr2: {}},
+		},
+		{
+			name: "two addresses all invalid predicates",
+			testTuple: []testTuple{
+				{address: addr1, isValidPredicate: false},
+				{address: addr2, isValidPredicate: false},
+				{address: addr1, isValidPredicate: false},
+				{address: addr1, isValidPredicate: false},
+			},
+			expectedRes: map[common.Address][]byte{addr1: {7}, addr2: {1}},
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			require := require.New(t)
+			// Create the rules from TestChainConfig and update the predicates based on the test params
+			rules := params.TestChainConfig.AvalancheRules(common.Big0, 0)
+			predicater := precompileconfig.NewMockPredicater(gomock.NewController(t))
+			predicater.EXPECT().PredicateGas(gomock.Any()).Return(uint64(0), nil).Times(len(test.testTuple))
+			validPredicateCount := 0
+
+			var txAccessList types.AccessList
+			for _, tuple := range test.testTuple {
+				predicateHash := invalidHash
+				if tuple.isValidPredicate {
+					validPredicateCount++
+					predicateHash = validHash
+				}
+				txAccessList = append(txAccessList, types.AccessTuple{
+					Address: tuple.address,
+					StorageKeys: []common.Hash{
+						predicateHash,
+					},
+				})
+			}
+
+			invalidPredicateCount := len(test.testTuple) - validPredicateCount
+			predicater.EXPECT().VerifyPredicate(gomock.Any(), validHash[:]).Return(true).Times(validPredicateCount)
+			predicater.EXPECT().VerifyPredicate(gomock.Any(), invalidHash[:]).Return(false).Times(invalidPredicateCount)
+			rules.Predicaters[addr1] = predicater
+			rules.Predicaters[addr2] = predicater
+
+			// Specify only the access list, since this test should not depend on any other values
+			tx := types.NewTx(&types.DynamicFeeTx{
+				AccessList: txAccessList,
+				Gas:        53000,
+			})
```

Comment:
Can we call these in order under in the if-else handling if `tuple.isValidPredicate`, so that it enforces the right ordering?

```suggestion
			var txAccessList types.AccessList
			for _, tuple := range test.testTuple {
				var predicateHash common.Hash
				if tuple.isValidPredicate {
					validPredicateCount++
					predicateHash = validHash
					predicater.EXPECT().VerifyPredicate(gomock.Any(), validHash[:]).Return(true)
				} else {
					predicateHash = invalidHash
					predicater.EXPECT().VerifyPredicate(gomock.Any(), invalidHash[:]).Return(false)
				}
				txAccessList = append(txAccessList, types.AccessTuple{
					Address: tuple.address,
					StorageKeys: []common.Hash{
						predicateHash,
					},
				})
			}

			rules.Predicaters[addr1] = predicater
			rules.Predicaters[addr2] = predicater

			// Specify only the access list, since this test should not depend on any other values
			tx := types.NewTx(&types.DynamicFeeTx{
				AccessList: txAccessList,
				Gas:        53000,
			})
```